node/service/src/tests.rs

```diff
@@ -436,34 +421,36 @@ fn run_specialized_test_w_harness<F: FnOnce() -> CaseVars>(case_var_provider: F)
 		// ancestor must match the chain derived one.
 		let highest_approved_ancestor_w_desc =
 			best_chain_containing_block.and_then(|best_chain_containing_block| {
-				let min_blocknumber = chain
-					.blocks_by_hash
-					.get(&best_chain_containing_block)
-					.map(|x| x.number)
-					.unwrap_or_default();
-				let highest_approved_ancestor_w_desc =
-					chain.highest_approved_ancestors(min_blocknumber, best_chain_containing_block);
-				if let (
-					Some(highest_approved_ancestor_w_desc),
-					Some(highest_approved_ancestor_block),
-				) = (&highest_approved_ancestor_w_desc, highest_approved_ancestor_block)
-				{
-					assert_eq!(
-					highest_approved_ancestor_block, highest_approved_ancestor_w_desc.hash,
-					"TestCaseIntegrity: Provided and expected approved ancestor hash mismatch: {:?} vs {:?}",
-					highest_approved_ancestor_block, highest_approved_ancestor_w_desc.hash,
-				);
-				}
-				highest_approved_ancestor_w_desc
-			});
-		if let Some(haacwd) = &highest_approved_ancestor_w_desc {
-			let expected = chain.undisputed_chain(haacwd.number, haacwd.hash);
-			assert_eq!(
-				expected, undisputed_chain,
-				"TestCaseIntegrity: Provided and anticipated undisputed chain mismatch: {:?} vs {:?}",
-				undisputed_chain, expected,
-			)
-		}
+				chain.blocks_by_hash.get(&target_block).map(|target_block_header| {
+                    let target_blocknumber = target_block_header.number;
+                    let highest_approved_ancestor_w_desc =
+                        chain.highest_approved_ancestors(target_blocknumber, best_chain_containing_block);
+                    if let (
+                        Some(highest_approved_ancestor_w_desc),
+                        Some(highest_approved_ancestor_block),
+                    ) = (&highest_approved_ancestor_w_desc, highest_approved_ancestor_block)
+                    {
+                        assert_eq!(
+                            highest_approved_ancestor_block, highest_approved_ancestor_w_desc.hash,
+                            "TestCaseIntegrity: Provided and expected approved ancestor hash mismatch: {:?} vs {:?}",
+                            highest_approved_ancestor_block, highest_approved_ancestor_w_desc.hash,
+                        );
+
+                        let expected = chain.undisputed_chain(
+                            target_blocknumber,
```

Comment:
Undisputed chain computes the parent of the first dispute after the target block. If there is no dispute after the target block, then `highest_approved_ancestor == expected`.